src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,308 @@
+            => FlushPacketAsync(true, isAsync: false, CancellationToken.None).Wait();
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken ct)
+            => await FlushPacketAsync(true, true, ct).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public void SetPacketSize(int bufferSize)
+        {
+            _writeBuffer = new byte[bufferSize];
+            InitializeHeaderBytesToZero();
+        }
+
+        /// <inheritdoc />
+        public void ReplaceUnderlyingStream(Stream stream) => _underlyingStream = stream;
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void Write(byte[] buffer, int offset, int count)
+            => Write(buffer.AsSpan(offset, count));
+
+
+        /// <inheritdoc/>
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            int len = buffer.Length;
+            // The buffer may not have enough space. Write what we can and then flush the buffer with a soft flush.
+            while (len > 0)
+            {
+                if (len > _writeBuffer.Length - _writeBufferEnd)
+                {
+                    // Only a part of the length fits in the buffer.
+                    int bytesToWrite = _writeBuffer.Length - _writeBufferEnd;
+                    buffer[..bytesToWrite].CopyTo(_writeBuffer.AsSpan(_writeBufferEnd));
+                    _writeBufferEnd += bytesToWrite;
+                    len -= bytesToWrite;
+                    FlushPacketAsync(false, isAsync: false, CancellationToken.None).ConfigureAwait(false);
+                }
+                else
+                {
+                    // The whole length can be added to the buffer.
+                    buffer.CopyTo(_writeBuffer.AsSpan(_writeBufferEnd));
+                    _writeBufferEnd += len;
+                    len = 0;
+                }
+            }
+        }
+
+        /// <summary>
+        /// An overload for writing a byte to the stream.
+        /// Use this in case the caller is always going to write asynchronously. 
+        /// Otherwise use the WriteByteAsync because it takes the isAsync flag.
+        /// </summary>
+        /// <param name="value">The byte value to be written to the stream.</param>
+        public override void WriteByte(byte value)
+        {
+            // If we are already at the end of the buffer, flush the buffer, with a softflush.
+            if (_writeBuffer.Length - _writeBufferEnd == 0)
+            {
+                FlushPacketAsync(false, isAsync: false, CancellationToken.None).Wait();
```

Comment:
No Waits should be used.
You should either write truly Sync API to FlushPacket or do not expose WriteByte if it cannot be made truly sync.
Only use Async APIs where needed, and GetAwaiter().GetResult() from parent caller will make the async API sync.